piptools/scripts/compile.py

```diff
@@ -498,10 +498,18 @@ def cli(
                 log.error(str(e))
                 log.error(f"Failed to parse {os.path.abspath(src_file)}")
                 sys.exit(2)
+            package_name = metadata.get_all("Name")[0]
             comes_from = f"{metadata.get_all('Name')[0]} ({src_file})"
             constraints.extend(
                 [
-                    install_req_from_line(req, comes_from=comes_from)
+                    install_req_from_line(
+                        req.replace(
+                            package_name, os.path.dirname(os.path.abspath(src_file))
+                        ),
+                        comes_from,
+                    )
+                    if package_name in req
```

Comment:
This search looks too broad. I wonder what if a package_name clashes with one of a dependency?

Reply:
Could you give an example so i can test it

Reply:
```toml
[project]
name = "est"
version = "0.0.1"
dependencies = ["pytest"]
[project.optional-dependencies]
dev = ["pytest[testing]"]
```

Reply:
Error:
```
pip._internal.exceptions.InstallationError: Invalid requirement: 'py/Users/albert/Projects/pip-tools/foo'
Hint: It looks like a path. File 'py/Users/albert/Projects/pip-tools/foo' does not exist.
```

Reply:
so, i should parse req line, get package name first

Reply:
I can get full name from parse_req_from_line().requirement.name, but parse_req_from_line would be called twice also using install_req_from_line